R/post-parsers.R

```diff
@@ -0,0 +1,185 @@
+#' Plumber Parsers
+#'
+#' Parsers are used in Plumber to transform the raw body content received
+#' by a request to the API.
+#' @name parsers
+#' @rdname parsers
+NULL
+
+#' Add a Parsers
+#'
+#' A parser is responsible for decoding the raw body content of a request into
+#' a list of arguments that can be mapped to endpoint function arguments.
+#' For instance, the \code{parser_json} parser content-type `application/json`.
+#' The list of available parsers in plumber is global.
+#'
+#' @param name The name of the parser (character string)
+#' @param parser The parser to be added.
+#' @param pattern A pattern to match against the content-type of each part of
+#' the request body.
+#'
+#' @details For instance, the \code{parser_json} pattern is `application/json`.
+#' If `pattern` is not provided, will be set to `application/{name}`.
+#' Detection is done assuming content-type starts with pattern and is
+#' case sensitive.
+#'
+#' Parser function structure is something like below. Available parameters
+#' to build parser are `value`, `content_type` and `filename` (only available
+#' in `multipart-form` body).
+#' ```r
+#' parser <- function(...) {
+#'   function(value, content_type = "ct", filename, ...) {
+#'     # do something with raw value
+#'   }
+#' }
+#' ```
+#'
+#' It should return a named list if you want values to map to
+#' plumber endpoint function args.
+#'
+#' @examples
+#' parser_json <- function(...) {
+#'   function(value, content_type = "application/json", ...) {
+#'     charset <- getCharacterSet(content_type)
+#'     value <- rawToChar(value)
+#'     Encoding(value) <- charset
+#'     safeFromJSON(value)
+#'   }
+#' }
+#' @md
+#' @export
+addParser <- function(name, parser, pattern = NULL) {
+  if (is.null(.globals$parsers)) {
+    .globals$parsers <- list()
+  }
+  if (!is.null(.globals$parsers$func[[name]])) {
+    stop("Already have a parser by the name of ", name)
+  }
+  if (is.null(pattern)) {
+    pattern <- paste0("application/", name)
+  }
+  .globals$parsers$func[[name]] <- parser
+  .globals$parsers$pattern[[name]] <- pattern
+}
+
+
+
+#' JSON
+#' @rdname parsers
+#' @param ... Raw values and headers are passed there.
+#' @export
+parser_json <- function(...) {
+  function(value, content_type = NULL, ...) {
```

Comment:
Could have a more informative argument name than `value` (maybe `postBodyRaw`)?

Reply:
reusing the argument name from `webutils::parse_multipart`. Otherwise it breaks `do.call(parser(), toparse)`. Or should I rename ouputs from `webutils::parse_multipart`.

Let me know and I'll change them.